File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    List<String>ans = new ArrayList<>();
    public List<String> letterCombinations(String digits) {
        if(digits.length()==0) return new ArrayList<>();
        HashMap<Character, List<Character>>map = new HashMap<>();
        map.put('2', List.of('a','b','c'));
        map.put('3', List.of('d','e','f'));
        map.put('4', List.of('g','h','i'));
        map.put('5', List.of('j','k','l'));
        map.put('6', List.of('m','n','o'));
        map.put('7', List.of('p','q','r','s'));
        map.put('8', List.of('t','u','v'));
        map.put('9', List.of('w','x','y','z'));
        getLetterCombinations(digits, 0, digits.length(), new StringBuilder(),map);
        return ans;
    }
    public void getLetterCombinations(String digits, int index, int n, 
    StringBuilder sb, HashMap<Character, List<Character>>map){
        if(sb.length() == n){
            ans.add(sb.toString());
            return;
        }
        List<Character>curList = map.get(digits.charAt(index));
        for(Character c: curList){
            sb.append(c);
            getLetterCombinations(digits, index+1, n, sb, map);
            sb.deleteCharAt(sb.length()-1);
        }
    }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string>ans;
    unordered_map<char,string>map={{'2',"abc"},{'3',"def"},{'4',"ghi"}
    ,{'5',"jkl"},{'6',"mno"},{'7',"pqrs"},{'8',"tuv"},{'9',"wxyz"}};
    vector<string> letterCombinations(string digits) {
        if(digits.length()==0) return ans;
        string s="";
        backtrack(digits,0,s);
        return ans;
    }
    void backtrack(string& digits,int index,string s){
        if(s.length()==digits.length()){
            ans.push_back(s);
            return;
        }
        string tmp=map[digits[index]];
        for(char c:tmp){
            s+=c;
            backtrack(digits,index+1,s);
            s.erase(s.length()-1);
        }
    }
};